File: leetcode/0797-all-paths-from-source-to-target/Solution.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

class Solution {
    private int target;
    private List<List<Integer>> results;
    private LinkedList<Integer> path;

    public List<List<Integer>> allPathsSourceTarget(int[][] graph) {
        target = graph.length - 1;
        results = new ArrayList<>();
        path = new LinkedList<>();
        path.add(0);
        dfs(graph, 0);
        return results;
    }

    private void dfs(int[][] graph, int node) {
        if (node == target) {
            results.add(new ArrayList<>(path));
            return;
        }
        for (int neighbor : graph[node]) {
            path.add(neighbor);
            dfs(graph, neighbor);
            path.removeLast();
        }
    }
}

File: leetcode/0797-all-paths-from-source-to-target/solution.py
from typing import List


class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        target = len(graph) - 1
        results = []
        path = [0]

        def dfs(node: int) -> None:
            if node == target:
                results.append(path[:])
                return
            for neighbor in graph[node]:
                path.append(neighbor)
                dfs(neighbor)
                path.pop()

        dfs(0)
        return results
